core/src/broadcast_stage/broadcast_duplicates_run.rs

```diff
@@ -269,44 +273,58 @@ impl BroadcastRun for BroadcastDuplicatesRun {
                 .map(|(pubkey, _)| *pubkey)
                 .collect()
         };
+
         // Broadcast data
         let cluster_nodes =
             self.cluster_nodes_cache
                 .get(slot, &root_bank, &working_bank, cluster_info);
         let socket_addr_space = cluster_info.socket_addr_space();
         let packets: Vec<_> = shreds
             .iter()
-            .filter_map(|shred| {
+            .flat_map(|shred| {
                 let seed = shred.seed(Some(self_pubkey), &root_bank);
-                let node = cluster_nodes.get_broadcast_peer(seed)?;
+                let node = cluster_nodes.get_broadcast_peer(seed);
+                if node.is_none() {
+                    return vec![];
+                }
+                let node = node.unwrap();
                 if !socket_addr_space.check(&node.tvu) {
-                    return None;
+                    return vec![];
                 }
-                if self.original_last_data_shreds.contains(&shred.signature()) {
-                    // If the node is within the partitin skip the shred.
+                if self
+                    .original_last_data_shreds
+                    .lock()
+                    .unwrap()
+                    .remove(&shred.signature())
+                {
                     if cluster_partition.contains(&node.id) {
                         info!(
                             "skipping node {} for original shred index {}, slot {}",
                             node.id,
                             shred.index(),
                             shred.slot()
                         );
-                        return None;
+                        return vec![];
                     }
+                } else if self
+                    .partition_last_data_shreds
+                    .lock()
+                    .unwrap()
+                    .remove(&shred.signature())
+                {
+                    // If the shred is part of the partition, broadcast it directly to
+                    // the partition node
+                    return cluster_partition
+                        .iter()
+                        .filter_map(|pubkey| {
+                            let tvu = cluster_info
+                                .lookup_contact_info(pubkey, |contact_info| contact_info.tvu)?;
+                            Some((&shred.payload, tvu))
+                        })
+                        .collect();
```

Comment:
Always broadcast partition shreds to the validators in the partition

Reply:
This seems to add an asymmetry between when it has to broadcast to the partition vs the non-partition?

Should this wide fanout be conditional on `self.config.stake_partition` being less than some threshold? e.g. `self.config.stake_partition < 10%` or something.
What if `get_broadcast_peer` already returns a node within the partition?

Also, not related to this specific change, but seems like this whole partitioning mechanism only happens during broadcast stage. The retransmit stage is not aware of the partition and will fanout to its subtree. So in the case that we stick with `get_broadcast_peer` (or, otherwise include the very first node in first neighborhood) it will fanout to the entire tree.
Is that fine for purpose of the tests? or, should this always do a wide fantout to all the target nodes and then not use retransmit-stage at all for the tests?